### ISSUE TITLE
Remove RTTI usage

### ISSUE THREAD

**Issue by brunoabinader** [mapbox/mapbox-gl-native#12576]
RTTI increases our binary footprint, and `dynamic_cast` can be substantially low in runtime under certain circumstances. We've removed our usage of `dynamic_cast` and `typeid` completely (see https://github.com/mapbox/mapbox-gl-native/pull/12455 and https://github.com/mapbox/mapbox-gl-native/pull/12475), however there is still some code that depends on Boost.Spirit headers which uses `typeid` internally. According to [this comment](https://github.com/mapbox/mapbox-gl-native/pull/12475#issuecomment-408921522) from @ChrisLoer, there are alternatives for that. Once we solve that, we're then clear to add `-fno-rtti` as part of the compiler parameters to disable RTTI.

/cc @mapbox/gl-core @mapbox/embedded

**Comment by kkaefer** [comment]
I did some digging and found that Boost Spirit doesn't use RTTI/`typeid` directly: it's only used by way of the Boost IOStreams library. There, it's not necessary either in Boost Spirit's usage of it; I just commented out both lines, and was able to compile Mapbox GL for Android with `-fno-rtti`. This saved ~60KB of the binary size.

Also see https://github.com/boostorg/iostreams/pull/68 for an alternative to commenting out the iostreams functions that use `typeid`.
